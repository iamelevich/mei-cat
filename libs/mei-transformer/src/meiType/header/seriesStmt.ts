import { Type } from "@sinclair/typebox";
import { StandardTagSchema } from "../common";
import { AttrBiblSchema } from "../shared/attr/bibl";
import { AttrCommonSchema } from "../shared/attr/common";

/**
 * Groups information about the series, if any, to which a publication belongs.
 * @see https://music-encoding.org/guidelines/v5/elements/seriesStmt.html
 */
export const SeriesStmtSchema = Type.Intersect([
	StandardTagSchema,
	AttrCommonSchema,
	AttrBiblSchema,
	Type.Object(
		{
			/**
			 * Contains any heading, for example, the title of a section of text, or the heading of a list.
			 * @see https://music-encoding.org/guidelines/v5/elements/head.html
			 */
			head: Type.Optional(
				Type.Union([Type.Ref("head"), Type.Array(Type.Ref("head"))]),
			),
			/**
			 * Title of a bibliographic entity.
			 * @see https://music-encoding.org/guidelines/v5/elements/title.html
			 */
			title: Type.Union([Type.Ref("title"), Type.Array(Type.Ref("title"))]),
			/**
			 * Transcription of text that names one or more individuals, groups, or in rare cases, mechanical processes, responsible for creation, realization, production, funding, or distribution of the intellectual or artistic content.
			 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
			 */
			respStmt: Type.Optional(
				Type.Union([Type.Ref("respStmt"), Type.Array(Type.Ref("respStmt"))]),
			),
			/**
			 * An alpha-numeric string that establishes the identity of the described material.
			 * @see https://music-encoding.org/guidelines/v5/elements/identifier.html
			 */
			identifier: Type.Optional(
				Type.Union([
					Type.Ref("identifier"),
					Type.Array(Type.Ref("identifier")),
				]),
			),
			/**
			 * Defines the scope of a bibliographic reference, for example as a list of page numbers, or a named subdivision of a larger work.
			 * @see https://music-encoding.org/guidelines/v5/elements/biblScope.html
			 */
			biblScope: Type.Optional(
				Type.Union([Type.Ref("biblScope"), Type.Array(Type.Ref("biblScope"))]),
			),
			/**
			 * List of the material contained within a resource.
			 * @see https://music-encoding.org/guidelines/v5/elements/contents.html
			 */
			contents: Type.Optional(
				Type.Union([Type.Ref("contents"), Type.Array(Type.Ref("contents"))]),
			),
			/**
			 * Groups information about the series, if any, to which a publication belongs.
			 * @see https://music-encoding.org/guidelines/v5/elements/seriesStmt.html
			 */
			seriesStmt: Type.Optional(
				Type.Union([
					Type.Ref("seriesStmt"),
					Type.Array(Type.Ref("seriesStmt")),
				]),
			),
		},
		{ additionalProperties: false },
	),
]);
