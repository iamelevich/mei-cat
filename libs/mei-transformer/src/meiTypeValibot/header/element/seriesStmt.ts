import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import {
	AttrBiblSchema,
	AttrCommonSchema,
	BiblScopeSchema,
	type BiblScopeData,
	ModelHeadLikeSchema,
	type ModelHeadLikeData,
	ModelIdentifierLikeSchema,
	type ModelIdentifierLikeData,
	ModelRespLikePartSchema,
	type ModelRespLikePartData,
	ModelTitleLikeSchema,
	type ModelTitleLikeData,
	RespStmtSchema,
	type RespStmtData,
} from "../../shared";
import { ContentsSchema, type ContentsData } from ".";

/**
 * Base schema with attribute, to simplify types for SeriesStmtSchema
 */
const SeriesStmtBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
});

type SeriesStmtBaseData = v.InferOutput<typeof SeriesStmtBaseSchema>;

/**
 * Groups information about the series, if any, to which a publication belongs.
 * @see https://music-encoding.org/guidelines/v5/elements/seriesStmt.html
 */
export const SeriesStmtSchema: v.GenericSchema<SeriesStmtData> = v.intersect([
	SeriesStmtBaseSchema,
	v.object({
		/**
		 * Reference to element biblScope
		 * @see https://music-encoding.org/guidelines/v5/elements/biblScope.html
		 */
		biblScope: v.optional(
			v.union([
				v.lazy(() => BiblScopeSchema),
				v.array(v.lazy(() => BiblScopeSchema)),
			]),
		),
		/**
		 * Reference to element contents
		 * @see https://music-encoding.org/guidelines/v5/elements/contents.html
		 */
		contents: v.optional(
			v.union([
				v.lazy(() => ContentsSchema),
				v.array(v.lazy(() => ContentsSchema)),
			]),
		),
		/**
		 * Reference to element respStmt
		 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
		 */
		respStmt: v.optional(
			v.union([
				v.lazy(() => RespStmtSchema),
				v.array(v.lazy(() => RespStmtSchema)),
			]),
		),
		/**
		 * Reference to element seriesStmt
		 * @see https://music-encoding.org/guidelines/v5/elements/seriesStmt.html
		 */
		seriesStmt: v.optional(
			v.union([
				v.lazy(() => SeriesStmtSchema),
				v.array(v.lazy(() => SeriesStmtSchema)),
			]),
		),
	}),
	ModelHeadLikeSchema,
	ModelIdentifierLikeSchema,
	ModelRespLikePartSchema,
	ModelTitleLikeSchema,
]);

export type SeriesStmtData = SeriesStmtBaseData & {
	biblScope?: BiblScopeData | BiblScopeData[];
	contents?: ContentsData | ContentsData[];
	respStmt?: RespStmtData | RespStmtData[];
	seriesStmt?: SeriesStmtData | SeriesStmtData[];
} & ModelHeadLikeData &
	ModelIdentifierLikeData &
	ModelRespLikePartData &
	ModelTitleLikeData;
