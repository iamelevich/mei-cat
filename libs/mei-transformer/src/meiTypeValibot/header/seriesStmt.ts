import * as v from "valibot";
import { StandardTagSchema } from "../common";
import { AttrCommonSchema } from "../shared/attr/common";
import { AttrLangSchema } from "../shared/attr/lang";

/**
 * Simplified schema for seriesStmt element.
 * @see https://music-encoding.org/guidelines/v5/elements/seriesStmt.html
 */
export const SeriesStmtSchema = v.object({
	...StandardTagSchema.entries,
	...AttrCommonSchema.entries,
	...AttrLangSchema.entries,
});

export type SeriesStmtData = v.InferOutput<typeof SeriesStmtSchema>;
