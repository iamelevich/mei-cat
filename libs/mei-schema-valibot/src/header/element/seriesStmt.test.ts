import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { SeriesStmtSchema } from "./seriesStmt";

describe("SeriesStmtSchema", () => {
	it("should be defined", () => {
		expect(SeriesStmtSchema).toBeDefined();
	});

	it("should be a valid schema", () => {
		const parseResult = v.parse(SeriesStmtSchema, {
			title: {
				"#text": "Carl Nielsen Works",
				"@xml:id": "title_adfefe6a",
			},
			identifier: {
				"#text": "CNW",
				"@type": "file_collection",
				"@xml:id": "identifier_47b68436",
			},
			"@xml:id": "seriesStmt_N202E5",
		});

		expect(parseResult).toMatchSnapshot();
	});
});
