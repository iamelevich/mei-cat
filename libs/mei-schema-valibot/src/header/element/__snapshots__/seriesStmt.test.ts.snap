// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`SeriesStmtSchema should be a valid schema 1`] = `
{
  "@xml:id": "seriesStmt_N202E5",
  "identifier": {
    "#text": "CNW",
    "@type": "file_collection",
    "@xml:id": "identifier_47b68436",
  },
  "title": {
    "#text": "Carl Nielsen Works",
    "@xml:id": "title_adfefe6a",
  },
}
`;
